### ISSUE TITLE
Pygame 2 Deprecation: Tracking stats

### ISSUE THREAD

**Issue by MyreMylar** [pygame/pygame#2326]
I mentioned a few times in various other issues related to things we can't do while we maintain support for python 2 that it would be useful to begin tracking download stats to given an idea of how many of pygame's users are still using python 2. After all, there would be no point us continuing to support python 2 if almost no-one was still using it.

I think a reasonable usage proxy for python 2 usage is downloads from PyPi (the place where packages come form when you do `pip install pygame`). Not everyone gets pygame this way, but a lot of our users do and probably a decent cross section as pip is available on every platform and seems to be widely used these days.

Luckily for us there is a site that specifically tracks pypi download stats already. And it also has nice graphs which we can look at and paste into this issue periodically. 

Here is the relevant URL:
https://pypistats.org/packages/pygame

And here is the relevant graph:

![image](https://user-images.githubusercontent.com/13382426/99096645-6d2ba680-25ce-11eb-984f-28aaf474b4af.png)

As you can see it looks like our current python 2 downloads are around 20% of the total. It looks like basically all python 2 downloads are 2.7 in case anyone was concerned about that. Python 2.7 is the second most used minor version of python (it competes with several well used minor versions of python 3).

Some of these downloads will be people, including us, downloading pygame on 2.7 for their CI tests (or just local testing  in my case) but that definitely doesn't account for all of them. I would propose that we don't think about dropping python 2 support until the graph above has python 2 usage hovering below 5% for at least three months.

I will try to post a new version of the graph in this thread once every three months (or more regularly if it seems interesting), but anyone else should feel free to update it as well. There are some changes afoot in the Linux and OSX worlds that should start to see python 2 installs declining in 2021 and beyond.

**Comment by ankith26** [comment]
Fun fact: A good number of python 2 pip installs are either typo (missing pip3 with pip) or newbies who don’t quite get the difference between pip and pip3 (I was in that category for a while :) I had done pip install of many stuff and they didn’t work, then I got to know)

**Comment by ankith26** [comment]
I guess the title should have been **python 2** deprecation??
**pygame 2 deprecation** kinda gives the impression that pygame 2 is getting deprecated

**Comment by MyreMylar** [comment]
Yes, that was the intended title. Oops.

**Comment by illume** [comment]
Good idea.

TLDR; My current gut feeling is that we look at it again end of next year, but python2 will be popular for four years. Linking python 2 related issues to here will help us make a better decision. There are some risks that **may** make maintainability harder next year, and there are very good reasons for using python 2 still to do with distribution.

Some random notes:
- Linking python 2 related issues like you've done to this one will help us make better decisions.
- According to that graph... python 2 usage is going up! My current theory is... python 3.9 dropped support for Windows 7. The mistake with this is that people seem confused, and turn to python 2 "oh, python 3 doesn't support windows 7". Note the increase in python 2 users VS before python 3.9. This is happening across multiple python projects. Django (for example) now has more 2.7 installs some days... and they stopped supporting python 2 in new releases some years ago. They made some small tweaks to the website, but I'm not convinced it will make a big enough effect. Microsoft has taken over even more of python now, so I don't see the windows 7 situation changing.
- keeping track of benefits we get for dropping python 2 support is also useful for making the decision of when. The main benefits I see are reduced maintenance burden, cleaner code (reduced ifdefs), being able to use newer C features, and newer python features.
- How do we quantify reduced maintenance? Is it a tax on every single feature/bug fix, or just rarely? My feeling taking a sampling of the previously merged PRs is that it's a very small tax - but it's there. From a sampling of 20 merged PRs, one had an issue with python 2 (it was a python 3 fix for some unwanted testutils code which was deleted). I did another sampling of 20 merged PRs and didn't see any python 2 related work. 
- We have 20 "if PY2" and 132 "if PY3". There are 237 "if IS_SDLv1" and 129 "if IS_SDLv2".
- Is the code too complex with python 2 support in there to make improving modules too hard? Maybe the lack of python2 related work is because python 2 makes that code to hard to modify? Take the surfarray module, and buffer related code for example. There are some improvements that could be made to that code, but would they be too hard with python 2 support required? I'm not sure about this.
- What are the benefits of new C features worth? Are we missing out on supporting python 3 features at the API level, or can they be added anyway and just not supported on python 2. An example of something which doesn't stop us is supporting pathlib.Path objects. Something that would be harder to add python 3 only is "async/await" support (ignoring greenlet fans for a moment who may interject here about async/await being bad anyway and we should use greenlets).
- The first linked issue with C dot notation is a nice-to-have in my opinion. By itself, it's not such a big thing... but if we collect 20-30 of these little things?
- mac is keeping python 2 for now, as are a number of linux distros. Macs usually last a long time compared to other brands (several years of usage is not uncommon).
- some python 2 based pygame apps get more downloads than all time downloads on pypi in a single **hour**. I say this just to point out that pypi download stats aren't entirely representative of the userbase, even though it's a useful signal.
- pygame has a very low amount of CI users compared to other projects. There's also lots of people who will use the built in python.
- python 2 supports windows XP, python 3 doesn't. XP has a 31% share* in China for example (worlds largest internet population). There's several years of laptop sales to catch up to the hundreds of millions of XP users. There's a lot of old incompatible hardware out in the world. I feel in education, the cheaper and second hand laptops are going to be a thing for a long time.
- user base of python 2 is one thing, percentage of python 2 code is another statistic. Many pygame apps are untested hobby projects, and upgrading python code isn't something people have time to do generally. What is the statistic of python 2 supporting code VS python 3 supporting code? Is it 75% of pygame using code at python2, or 50%? I did a random sampling of 20 github projects, and 16/20 were python 3. I didn't bother with the pygame website, but I guess it's a bit more skewed towards python 2 with all the ancient ones on there that haven't been updated in years.
- Why 5% of users and not another number? If python 2 users drop to 5% of users on pypi stats, that's still higher than mac users. Do we drop mac support now? What if downloads overall increase 3x? Do we then drop python 2 support if it's (5/3)% of users? What if the number of mobile users increases dramatically, and all of them are on python 3.8?
- Build tooling is probably going to be the thing that hurts most. Already some people in the python packaging world have been trying their hardest to kill off python 2 support for years. I wouldn't be surprised if next year there might be some difficulties with some components like setuptools, pip, cibuildwheels and such. This was the case with python 3.4... the tooling dropped support for it, so it became quite a lot harder to support. When pip stopped working on the main version of Raspberry pi 1.5 years ago, they basically pinned pip to the version that worked, so this is probably a solution.
- python 2 supports better single file exe generation on windows. Which I think is a very good reason for people shipping apps on windows to use it. There was at least one pypy developer interested in supporting this use case. From an app distribution perspective, python 2 is better on Mac as well... since that's the python every mac user has. 
- python 2 is still slightly faster for many pygame apps than python 3.8 (see pygame.examples.testsprite benchmarks). Audio/IOT/video apps are another one where there is an advantage to python 2 due to less latency jitter. It could be that python 3.9 is faster if we tweak some API calls to use the new faster way of parsing arguments. For one user, an instrument simply does not work with python 3.8. The fps is too slow, and the timing is a little bit off. However, I expect some optimizations to their app, or pygame could tip the difference.
- What about book/tutorial/course percentage? Some of the most popular pygame books are still on python 2, but most (not all) in the last few years have been on python 3 or 2+3. What I've seen is university courses are mostly upgraded to python 3 years ago. However, I just saw one a few days ago which only just upgraded to python 3. The most popular pygame youtube videos are all using python 3.

**Comment by MyreMylar** [comment]
Those are definitely all good things to consider. 

I'm personally in no great hurry to push python 2 out of the door as I have mostly adapted to the quirks when developing with it. Like moving all your variable definitions to the top of a function, even though this sometimes this makes for slightly odd code where you need a variable - but don't necessarily need it in all paths through a function so you either end up making lots of sub functions for the sake of python 2, or just defining some variables that will go unused. Anyway, I've mostly adapted and it trips me up less.

I just wanted a place to track some data on python 2 usage for those times when we have to justify why we are supporting python 2 to issue reporters and contributors we ask to jump through python 2 related hoops, who reasonably ask 'Wasn't that dropped by python themselves at the start of the year?'.

One thing we should also remember too though is that even if we do drop python 2 support from future versions, users of python 2.7 will still have access to all the versions of pygame up to that point for all time. It's not like python 3 where if we stop releasing pygame will go stale because of the new versions of python that come out in the meantime. 

I wonder if PyPi/pip be made to grab the most recent binary version available for the requesters version of python. So Python 2.7 users doing `pip install pygame` would get the last version where we stopped python 2 support, and python 3 users would continue to get the latest version. 

I looked up the OSX 11 situation after the release this week and it looks like Apple has chickened out of dropping python 2 as their default python, as it was reported they were going to last year, so probably python 2 is going to be around for a good while yet. Maybe they will finally ditch it next year.

**Comment by illume** [comment]
> I wonder if PyPi/pip be made to grab the most recent binary version available for the requesters version of python. So Python 2.7 users doing pip install pygame would get the last version where we stopped python 2 support, and python 3 users would continue to get the latest version.

Seems to grab an old numpy for python2 for me.

**Comment by Starbuck5** [comment]
I'm in favor of dropping python 2 support, for code cleanliness and maintainability's sake mostly.

I don't think dropping it will have much adverse impact. Most/many pygame users still are on pygame 1.9.6 (qualitative observation from support prompt from exe's I played in pyweek31), so I doubt most of our python 2 users would even notice the change. Especially because if they did `pip install pygame --upgrade`, it would just upgrade to our last python 2 supporting version.

> Build tooling is probably going to be the thing that hurts most.

@ankith26 had trouble with getting Github actions to download an old enough compiler for python2 in #2507 (correct me if I'm wrong, but I think we had a conversation where you told me this on discord).

> XP has a 31% share* in China for example

[This website](https://gs.statcounter.com/windows-version-market-share/desktop/china) puts it at only 2.48%

> If python 2 users drop to 5% of users on pypi stats, that's still higher than mac users. Do we drop mac support now?

Cross platform support is a great pygame selling point. Cross version support is less valuable to any given user writing their code in any given version.

I don't think a drop to python 2 support would hurt as badly as we seem to be assuming.
- Old pygame games that run in python 2 would still run, since pip on python2 would grab the last pygame version that supports it.
- Users who find out their pygame version doesn't support `XYZ` future feature won't be able to begrudge our non support of a EOL version too much.

Most importantly, regardless of when we drop Python 2, I think we should set a specific time or pygame version, rather than wait forever for a percentage to go down.

Possible times to drop Python 2:
- Jan 1 2022 (two extra years of support, we did good)
- pygame 2.1 (we can focus on _sdl2 features without the added tax of python2 concerns)
- pygame 2.2 (the only thing in this milestone so far is removing deprecated code, so why not remove deprecated python version too?)

**Comment by ankith26** [comment]
Well, I would love to drop python2 support too.

I feel that dropping python2 does not have much of a negative impact, as one should not be using python2 anyways, and even if they are on python2, they can always use the latest pygame version that supports python2.

Secondly, dropping python2 will certainly help maintainability. Supporting python2 means having to write a different codepath sometimes, having to deal with extra bugs and just more work in general. Now it is true that in most cases, python2 related issues pop up less frequently, but occationally they do pop up, and when they do, they are annoying to fix most of the times.
For instance, I had python2 related issues on 2 of my recent PRs, and they sure were irritating to fix.

Thirdly, Python2 support (and SDL1 support) removal will make the codebase more clean, and thus easier to maintain. It's kinda hard to read code when there are so many `#ifdef` sprinkled all across.

So I too believe that we should be getting rid of python2 and SDL1 support by pygame 2.1. And pygame 2.1 is still a bit of a long way from now, so I think we will have given more than enough time for python2 users to upgrade.

**Comment by illume** [comment]
Yes, the maintainability reasons are there. Removing py2 and sdl1 would make the codebase more joyful to work on :)

Python 2 usage is still significant according to the pypi pygame stats. The peaks hover between 20-40%. Not too different from when we last looked in November 2020.

![pygame-stats](https://user-images.githubusercontent.com/9541/114310039-b0c40700-9abf-11eb-8bb8-df32775287f2.png)

> So I too believe that we should be getting rid of python2 and SDL1 support by pygame 2.1.

I agree with the 2.1 release dropping SDL1 and python 2 support. What defines pygame 2.1 otherwise? I'd suggest we work through all the compatibility issues left on the pygame 2 list. Which doesn't necessarily mean we fix them all, but at least address them all and give it a good go. In other words - the list we have for milestone "2.0" should be addressed: https://github.com/pygame/pygame/milestone/4 either fixed, or moved out of the milestone because they are too hard/impossible/decided not worth it. Probably a few 2.0.x releases more to work through those issues (and other compatibility ones that come up).

Sound like a plan?

**Comment by illume** [comment]
On the other hand... if someone is enthusiastic to rip out py2 and sdl1 now... fixing those compatibility issues can be done with unit tests and using a separate install of pygame 1.9.x. Thinking about how I would fix those issues, having SDL1 support in there isn't really necessary or much more convenient.

I changed my mind to vote for ripping out SDL1 and PY2 now. haha.

**Comment by ankith26** [comment]
sounds like a really good plan! :)
I could take this one up, ripping all SDL1 and python2 related stuff from the source. But imma probably start working on this after a while, because we could still do 2-3 more 2.0.x releases with the current setup for SDL1 and Python2 support, and then take on this task

**Comment by llindstrom** [comment]
As the person who put in most of those `#if PY3` and `#if IS_SDLv2` conditionals as well as write compat.py I am all for ripping out the Py2/SDL1 code. I am willing to do my share, but if we want to split the workload someone can do the C code while someone else does the Python. We can merge the changes back into main whenever any Python 2 related bug fixes are completed.

**Comment by robertpfeiffer** [comment]
Although I am very in favour of ripping out Py2 compat soon(tm), I would like to point out that native wayland support is still not a thing. Before we rip out py2, we should probably release one version that works with wayland natively (because https://github.com/pygame/pygame/issues/2487) and supports proper server-side window decorations and the like.

**Comment by Starbuck5** [comment]
Let's wait until 2.1.dev1 to start taking PY2 and SDL1 compatibility out of the code.

I think it would be good to release 2.0.2 now, since it has a couple months of improvements over 2.0.1. And there is definitely room for more "2.0 bugfix" releases after that.

It sounds like lots of people are enthusiastic to take a crack at it, so when the time comes we could split up responsibilities by module.

> As the person who put in most of those #if PY3 and #if IS_SDLv2 conditionals as well as write compat.py I am all for ripping out the Py2/SDL1 code

Best endorsement :heart:

> we should probably release one version that works with wayland natively (because #2487) and supports proper server-side window decorations and the like.

What does this entail? I see there's an issue, #2529, but is it just a wheels thing or does pygame need code changes?

**Comment by robertpfeiffer** [comment]
> What does this entail? I see there's an issue, #2529, but is it just a wheels thing or does pygame need code changes?

It's *mostly* a wheel thing. I can build a SDL2 with wayland support and make a wheel. There are some open questions though:

- Will this break on computers without wayland?
- Is current manylinux too old to be a proper wayland build env?
- How do we test this when there are so many differences between gnome, KDE, sway, and weston?

There are also actual code changes needed, like 
- using server-side window decorations by default
- exposing an the API for decorations/implementing pygame borderless windows as client-side decorations
- warning when running on XWayland (yes, really. It's the worst of both worlds)

but these can wait, mostly. I would like the nest release to be installable via pip, so it **just works** properly on Debian, maybe with X11 by default, but with wayland driver available.

**Comment by nthykier** [comment]
> Never mind that Debian stable is still shipping a python2.7 version of pygame 1.9.4. Debian _testing_ is currently in an _absurd_ configuration where they ship a wayland-based desktop environment together with the SDL1.2-based PyGame 1.9.6 and Python 3.9. It will be quite some time before _our_ dropped support for Python 2.7 can make its way into Debian, and they have different packages for python2 and python3 anyway.

Debian released Bullseye recently (after your message), so now we are looking at a Python3 SDL1.2-based PyGame 1,9.6 from the Debian side. Not sure if that changes anything for you though.

**Comment by illume** [comment]
This graph is weird. Python 2 usage really dropped for a while... then came back up.
But not as high as it was, and it appears to be going down again. At least it's not growing like it was! That was weird.

<img width="1017" alt="Screenshot 2021-10-09 at 08 47 07" src="https://user-images.githubusercontent.com/9541/136647375-73be7181-2652-496c-a253-6c2386c31dfb.png">

**Comment by illume** [comment]
I'll write to the mailing list about dropping support in our next release. Meaning 2.0.2 will be our last sdl1/py2 release (or maybe a 2.0.3 if we make a big mistake). But I guess everyone here is on board for dropping py2/sdl1 after 2.0.2 (which should come out very soon).

**Comment by takluyver** [comment]
> I wonder if PyPi/pip be made to grab the most recent binary version available for the requesters version of python. So Python 2.7 users doing pip install pygame would get the last version where we stopped python 2 support, and python 3 users would continue to get the latest version.

This already works, so long as the package metadata describes the minimum compatible Python. This is the `python_requires` keyword in `setup.py`. [Here's an example](https://github.com/python-pillow/Pillow/blob/2d7c487cf8fbee80428248138d5632a5da1758f9/setup.py#L1017) of it in use (in Pillow). The check was added in pip 9.0, which came out almost 5 years ago.

It's worth taking care to do this at the same time as dropping Python 2 support. If you make a release which breaks Python 2 but doesn't have the metadata saying that, it's a pain to fix after the fact.

**Comment by MyreMylar** [comment]
> This graph is weird. Python 2 usage really dropped for a while... then came back up. But not as high as it was, and it appears to be going down again. At least it's not growing like it was! That was weird.
> 

Looks like it is tracking school summer holidays to me (installs dropping over July and August), which would make sense with python & pygame being widely used in education and education frequently being on older hardware. I know I download pygame a lot more often during school term times, though obviously I use python 3.

**Comment by takluyver** [comment]
The timing does look like it could relate to UK school holidays, but it looks like it's driven by higher numbers on Python 3, not so much lower ones on Python 2:

![image](https://user-images.githubusercontent.com/327925/136690357-601d4fff-a194-47f8-875c-744555e64582.png)

Maybe there were some holiday clubs doing things with pygame on Python 3? The download numbers for Linux are up in a similar pattern, so I wondered about something with Raspberry Pi - but that probably wouldn't show up on PyPI, because it would get wheels from https://www.piwheels.org/ . It would have to be a massively popular activity to be adding an extra ~50k downloads per day, though...

Maybe the timing with the summer holidays is just a coincidence, and there was some change in a big CI platform or in pip's caching behaviour that drove this shift. :confused:

**Comment by ankith26** [comment]
So 2.1 finally released and dropped py2/sdl1/py3.5 🥳 
IG this issue can be closed (more work needs to be done for py2/sdl1 drop, but that's being tracked in other issues ;) )
